workflows/rnaseq/downstream/helpers.Rmd

```diff
@@ -725,4 +726,37 @@ write.clusterprofiler.results <- function(res, cprof.folder, label){
     write.table(res.split, file=filename.split, sep='\t', quote=FALSE, row.names=FALSE)
     return(list(orig=filename.orig, split=filename.split))
 }
+
+
+#' Remove batch effect
+#' 
+#' @param x Numeric matrix containing log-expression values for a series of samples
+#' @param batches Vector of batch factors/vectors
+#' @param covariates Matrix or vector of numeric covariates to be adjusted for
+#' @param design Design matrix relating to treatment conditions to be preserved, usually the design matrix with all experimental factors other than the batch effects
+#' @param ... Other arguments are passed to lmFit
+removebatchEffect <-
+function(x, batches=c(), covariantes=NULL, design=matrix(1,ncol(x),1),...)
+{
+    if(length(batches) == 0 and is.null(covariates)){
+        return(as.matrix(x))
+    }
+    batches_processed <- c()
+    if(length(batches) != 0){
+        for(batch in batches){
+            batch <- as.factor(batch)
+            contrasts(batch) <- contr.sum(levels(batch))
+            batch <- model.matrix(~batch)[,-1,drop=FALSE]
+            batches_processed <- append(batches_processed, batch)
+        }
+    }
+    if(!is.null(covariates)) {
+        covariates <- as.matrix(covariates)
+    }
+    X.batch <- cbind(batches_processed, covariates)
+    fit <- lmFit(x, cbind(design, X.batch),...)
+    beta <- fit$coefficients[,-(1:ncol(design)),drop=FALSE]
+    beta[is.na(beta)] <- 0
+    as.matrix(x) - beta %*% t(X.batch)
+}
```

Comment:
I think this would benefit from a small test -- we wouldn't need to mock up a bunch of RNA-seq data for this, we could just use a small matrix.

And maybe as a temporary test you could copy the function over to a real-world RNA-seq analysis and give it a shot there -- no need to wait for a merge for that.